### PR TITLE
Add `slotIdTime :: SlotId -> UTCTime` to WalletLayer

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#465 

# Overview

- [x] Added a naive `slotIdTime` to the `WalletLayer` (agnostic to changing blockchain parameters)
- [x] Moved `waitForWalletSync` to WalletLayer
- [x] Duplicated http-bridge `Cardano.WalletSpec` to Jormungandr, and added a test that the wallet tip, once synced, corresponds to a time in `(now - slotLength, now)`

# Comments

⚠️ also includes #572, but the base branch is still master ⚠️

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -357,13 +360,15 @@ newWalletLayer
     :: forall s t. (Buildable (Tx t))
     => Trace IO Text
     -> Block (Tx t)
-        -- ^ Very first block
+        -- ^ Very first block (block0)
+    -> UTCTime
```
Comment (rvl) [inline lib/core/src/Cardano/Wallet.hs]: See also #527 because this is also adding config parameters to the wallet layer.

File: lib/http-bridge/test/integration/Cardano/WalletSpec.hs [hunk 2]
```diff
@@ -62,6 +64,22 @@ spec = do
                 unsafeRunExceptT (readWallet wallet wid)
             unless (tip > (SlotId 0 0)) $
                 expectationFailure ("The wallet tip is still " ++ show tip)
+
+        it "The time of the tip is recent" $ \(_, wallet) -> do
```
Comment (rvl) [inline lib/http-bridge/test/integration/Cardano/WalletSpec.hs]: I'm not sure whether "The time of the tip is recent" is true in general. The network can miss blocks for example, or the node could be temporarily disconnected from its peers.

**Review by rvl — COMMENTED** [review]
:+1:

File: exe/wallet/jormungandr/Main.hs [hunk 3]
```diff
@@ -352,7 +352,7 @@ cmdServe = command "serve" $ info (helper <*> cmd) $ mempty
                     handleNetworkUnreachable tracer
                 Left (ErrGetBlockchainParamsGenesisNotFound _) ->
                     handleGenesisNotFound (sb, tracer)
-                Left (ErrGetBlockchainParamsNoInitialPolicy _) ->
+                Left (ErrGetBlockchainParamsIncompleteParams _) ->
```
Comment (paweljakubas) [inline exe/wallet/jormungandr/Main.hs]: indeed, better error name

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -720,6 +732,18 @@ newWalletLayer tracer bp db nw tl = do
...
+        :: SlotId
+        -> UTCTime
+    _slotIdTime slN =
+        addUTCTime (convert $ slotLength * slots) block0Date
+      where
+        slots = fromIntegral . fromEnum $ (flatSlot slN) - (flatSlot sl0)
+        sl0 = block0 ^. #header ^. #slotId
+
+        convert :: DiffTime -> NominalDiffTime
+        convert = toEnum . fromEnum
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet.hs:763]: we can also use the fact that that `NominalDiffTime` and `DiffTime` both have `RealFrac` instances. Hence they have both `Real` and `Fractional` instances. If so we can ,
```
convert :: DiffTime -> NominalDiffTime
convert = fromRational . toRational
```

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 5]
```diff
@@ -705,7 +705,7 @@ fromFlatSlot n = SlotId e (fromIntegral s)
     s = n `mod` epochLength
 
 epochLength :: Integral a => a
-epochLength = 21600
+epochLength = 500
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs]: why this change?

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs [hunk 6]
```diff
@@ -158,6 +161,14 @@ block0 = Block
     , transactions = []
     }
 
+-- | Hard-coded block0 date for Byron
+--
+-- Corresponds to 2017-09-17T04:08:52+00:00.
+-- Retrieved from https://github.com/input-output-hk/cardano-sl/blob/403e26a1d33623f4537517535b7de37773ae9259/scripts/prepare-genesis/README.md#31-genesis-json-format
+-- and also matches 'lib/mainnet-genesis.json' in cardano-sl.
+byronBlock0Date :: UTCTime
+byronBlock0Date = posixSecondsToUTCTime 1505621332
```
Comment (paweljakubas) [inline lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs]: :+1:

**Review by paweljakubas — APPROVED** [review]
very nice PR, I am fond of tests added! You are very close to nail it :-) (some more places needs revision, you need to import `block0Date` in `lib/core/test/unit/Cardano/WalletSpec.hs` etc)

**Review by paweljakubas — CHANGES_REQUESTED** [review]
probably some more places will need updating of new `BlockchainParameters` definition:
```
[pawel@arch cardano-wallet]$ ag BlockchainParameters
lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs
41:    ( BlockchainParameters (..) )
176:    , getInitialBlockchainParameters
178:        -> ExceptT ErrGetBlockchainParams m (BlockchainParameters (Jormungandr n))
241:    , getInitialBlockchainParameters = \block0 -> do
270:                return $ BlockchainParameters

lib/jormungandr/test/integration/Main.hs
30:    ( BlockchainParameters (..), newWalletLayer )
207:    -> IO (NetworkLayer (Jormungandr n) IO, BlockchainParameters (Jormungandr n))
214:        getInitialBlockchainParameters jormungandr (coerce block0H)

lib/core/src/Cardano/Wallet.hs
24:    , BlockchainParameters (..)
368:data BlockchainParameters t = BlockchainParameters
383:    -> BlockchainParameters t
410:    BlockchainParameters block0 feePolicy (SlotLength slotLength) txMaxSize = bp

lib/core/test/unit/Cardano/WalletSpec.hs
20:    ( BlockchainParameters (..)
375:    let bp = BlockchainParameters block0 policy slotLength txMaxSize

lib/http-bridge/test/bench/Main.hs
19:    ( BlockchainParameters (..), WalletLayer (..), newWalletLayer )
215:    let bp = BlockchainParameters

lib/http-bridge/test/integration/Cardano/WalletSpec.hs
16:    ( BlockchainParameters (..), WalletLayer (..), newWalletLayer )
84:        let bp = BlockchainParameters block0 byronFeePolicy byronSlotLength byronTxMaxSize

lib/http-bridge/test/integration/Main.hs
23:    ( BlockchainParameters (..), newWalletLayer )
267:            let bp = BlockchainParameters

exe/wallet/http-bridge/Main.hs
60:    ( BlockchainParameters (..), WalletLayer )
285:            -> IO (NetworkLayer t IO, BlockchainParameters t)
290:            let bp = BlockchainParameters

exe/wallet/jormungandr/Main.hs
63:    ( BlockchainParameters (..), WalletLayer )
77:    ( ErrGetBlockchainParams (..), getInitialBlockchainParameters )
340:            -> IO (NetworkLayer t IO, BlockchainParameters t)
349:                runExceptT (getInitialBlockchainParameters jor (coerce block0H)) >>= \case
```

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 5 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: This was temporary. Jormungandr uses `500`, so without this the "slotIdTime of tip is recent"-test on jormungandr didn't pass. So we need to at least use a dynamic `SlotsPerEpoch`/`epochLength` from `BlockchainParameters` 🤔

File: lib/core/src/Cardano/Wallet.hs [hunk 7]
```diff
@@ -720,6 +738,21 @@ newWalletLayer tracer bp db nw tl = do
...
             DB.putCheckpoint db (PrimaryKey wid) (newPending (tx, meta) w)
 
+    _slotIdTime
+        :: SlotId
+        -> UTCTime
+    _slotIdTime slN =
+        addUTCTime (convert $ slotLength * slots) block0Date
+      where
+        slots = fromIntegral . fromEnum $ (flatSlot' slN) - (flatSlot' sl0)
+        sl0 = block0 ^. #header ^. #slotId
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: I don't think this should live in the `WalletLayer`. This is a pure function that could live in the `Primitive.Types` module.

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs (hunk 6 above)
Comment (KtorZ) [reply lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs]: Hmmmm.... This depends on the type of network, doesn't it? Also, this time seems wrong for mainnet anyway :s 
See also: 

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/http-bridge/test/bench/Main.hs#L323-L326

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs [hunk 8]
```diff
@@ -169,3 +183,14 @@ byronSlotLength = SlotLength $ secondsToDiffTime 20
...
+
+byronBlockchainParameters :: BlockchainParameters (HttpBridge n)
+byronBlockchainParameters = BlockchainParameters
+                    { getGenesisBlock = block0
+                    , getGenesisBlockDate = byronBlock0Date
+                    , getFeePolicy = byronFeePolicy
+                    , getSlotLength = byronSlotLength
+                    , getTxMaxSize = byronTxMaxSize
+                    , getSlotsPerEpoch = Quantity 21600
+                    }
```
Comment (KtorZ) [inline lib/http-bridge/src/Cardano/Wallet/HttpBridge/Compatibility.hs]: That's a lot of indentation somehow :sweat_smile:

File: lib/jormungandr/test/integration/Cardano/WalletSpec.hs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        let dir = "test/data/jormungandr"
+        handle <- async $ void $ launch
+            [ Command "jormungandr"
+                [ "--genesis-block", dir ++ "/block0.bin"
+                , "--config", dir ++ "/config.yaml"
+                , "--secret", dir ++ "/secret.yaml"
+                ] (return ())
+                Inherit
+            ]
+        threadDelay 1000000
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Cardano/WalletSpec.hs]:
```suggestion
threadDelay (1 * second)
```

File: lib/jormungandr/test/integration/Cardano/WalletSpec.hs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+        it "The time of the tip is recent" $ \(_, wallet) -> do
+            bytes <- entropyToBytes <$> genEntropy @(EntropySize 15)
+            let xprv = generateKeyFromSeed (Passphrase bytes, mempty) mempty
+            wid <- unsafeRunExceptT $ createWallet wallet
+                (WalletId $ digest $ publicKey xprv)
+                (WalletName "My Wallet")
+                (mkSeqState @(Jormungandr 'Testnet) (xprv, mempty) minBound)
+            unsafeRunExceptT $ restoreWallet wallet wid
+            threadDelay 8000000
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Cardano/WalletSpec.hs]:
```suggestion
threadDelay (8 * second)
```

File: lib/jormungandr/test/integration/Cardano/WalletSpec.hs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            tip <- slotId . currentTip . fst <$>
+                unsafeRunExceptT (readWallet wallet wid)
+            let mintedTime = slotIdTime wallet tip
+            t <- getCurrentTime
+            if t > (addUTCTime (-20) mintedTime)
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Cardano/WalletSpec.hs]: This number looks quite arbitrary, what's the rationale here?

Comment (Anviking) [reply lib/jormungandr/test/integration/Cardano/WalletSpec.hs]: The maximum delay should be `SlotDuration = 10 s`, so instead making sure the tip was minted within 20 seconds ago is just for some additional marginal.

Comment (KtorZ) [reply lib/jormungandr/test/integration/Cardano/WalletSpec.hs]: Would be good to make this explicit. Especially because the slot duration isn't a "global constant" that necessarily has the same value. We know it's `10` because we set it to `10` in the genesis block. The best would be to expose a `getBlockchainParameters` from the wallet layer and use the slot duration provided there.

Also, if the maximum delay is the slot duration, we should check for the slot duration, and not twice of it.

File: lib/core/src/Cardano/Wallet.hs (hunk 7 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet.hs]: Tbh, I'm thinking it might be best to go for the full solution (with dynamic slotting) straight away.

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs]: Seems reasonable. You don't need to make many assumption about how the slot times are determined in order to implement this function. You can expect an index mapping `SlotNo` (or `(Epoch, SlotId)` to a given `(SlotLength, EpochLength)` and make sure to correctly compute the corresponding time. 

I'd even expect this to be so isolated that it could live in its own PR with dedicated testing. No need to do any modification to the wallet layer to implement this.

File: lib/jormungandr/test/integration/Cardano/WalletSpec.hs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            bytes <- entropyToBytes <$> genEntropy @(EntropySize 15)
+            let xprv = generateKeyFromSeed (Passphrase bytes, mempty) mempty
+            wid <- unsafeRunExceptT $ createWallet wallet
+                (WalletId $ digest $ publicKey xprv)
+                (WalletName "My Wallet")
+                (mkSeqState @(Jormungandr 'Testnet) (xprv, mempty) minBound)
+            unsafeRunExceptT $ restoreWallet wallet wid
+            threadDelay (30 * second)
+            -- TODO: This wait time is curcial to allow the wallet to sync,
+            -- but it might be unrealistic to do so.
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Cardano/WalletSpec.hs]: Note that you could also wait for the wallet to be synced by polling it every second or so :) 
That should take far less than 30s on Jörmungandr actually.

Comment (Anviking) [reply lib/jormungandr/test/integration/Cardano/WalletSpec.hs]: 👍 Thanks! --> #572

File: lib/core/src/Cardano/Wallet.hs (hunk 7 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet.hs]: Or no, I changed my mind 😅 .

Storing a list, `[SlottingConfiguration]`,
```haskell
data SlottingConfiguration 
   { validFrom :: SlotId
    , slotLength :: SlotLength
    , epochLength :: EpochLength
    }
```
would be kind of plausible assuming the slotting configuration doesn't change regularly, but for _me_, the same fears as here https://github.com/input-output-hk/cardano-wallet/pull/563#discussion_r304824349 would apply.

**Comment by Anviking** [comment]
@rvl and I agreed that that we should keep the calculation naive for now. 

I also agreed to put `slotIdTime` in `NetworkLayer`, but I am having second thoughts about that. Currently, `JörmungandrLayer` allows fetching the initial blockchain parameters. HttpBridge does not. It would be weird to hard-code blockchainParameters inside the networkLayer, and it would be equally weird to take it as an argument. And none of `JörmungandrLayer`, `HttpBridgeLayer`, `NetworkLayer` has any storage.

@KtorZ suggested making `slotIdTime` a pure function in `Primitive.Types`. Unfortunately, we cannot then take `BlockchainParameters` as an argument since `BlockchainParameters` is defined in `Cardano.Wallet` and… import cycles…

I'd suggest keeping the naive `slotIdTime` inside `WalletLayer` as is for now. Or alternatively still in `Cardano.Wallet` but outside the `WalletLayer`.

File: lib/jormungandr/test/integration/Cardano/WalletSpec.hs [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+                (mkSeqState @(Jormungandr 'Testnet) (xprv, mempty) minBound)
+            unsafeRunExceptT $ restoreWallet wallet wid
+            threadDelay (8 * second)
+            tip <- slotId . currentTip . fst <$>
+                unsafeRunExceptT (readWallet wallet wid)
+            unless (tip > (SlotId 0 0)) $
+                expectationFailure ("The wallet tip is still " ++ show tip)
+
+        -- This tests both that the wallet syncs correctly, and that
+        -- 'slotIdTime' computes the time of the wallet tip correctly.
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Cardano/WalletSpec.hs:87]: So this test is testing two things… We could just get the tip from the `NetworkLayer` and call `slotIdTime` on it. We could get by without any `waitForWalletSync`. On the other hand I thought it was kind of interesting to use the `currentTip`.

**Comment by rvl** [comment]
> I'd suggest keeping the naive `slotIdTime` inside `WalletLayer` as is for now. Or alternatively still in `Cardano.Wallet` but outside the `WalletLayer`.

@Anviking As far as I can tell, at present, the only place we will need `slotIdTime` is in `_listTransactions`. So the function doesn't need to be a public method of `WalletLayer`. It can just be an auxiliary function of `_listTransactions` (where it will have slot duration and epoch length variables in scope).

**Comment by Anviking** [comment]
I'm not sure how we'd test that. I guess the approach would be to construct blocks with txs and specific slotIds, and somehow feed them to the wallet layer…

This sounds kind of tricky, but on the other hand it's testing more things. And if we decide to store every `UTCTime`  the test would be unchanged. 🤔

**Comment by KtorZ** [comment]
Superseded by #542
